File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # split list into 2 halves
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        # second half starts at slow
        second = slow.next
        # reverse the 2nd half of list
        prev = None
        slow.next = prev
        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp
        # reorder the list by traversing 1st list and 2nd list alternating between
        first = head
        second = prev
        while second:
            tmp1 = first.next
            tmp2 = second.next
            first.next = second
            second.next = tmp1
            first = tmp1
            second = tmp2
        

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* shadow = head;
        int len = 0;
        while(shadow) {
            shadow = shadow->next;
            len++;
            cout << len << endl;
        }
        if(len == 1) {
            return nullptr;
        }
        shadow = head;
        if(len - n == 0) {
            cout << "length is 0" << endl;
            cout << shadow->val;
            return head->next;
        }
        cout << "Shadow: " << shadow->val << len-n << endl;
        for(int i = 1; i < len - n; i++) {
            shadow = shadow->next;
        }
        cout << "Shadow: " << shadow->val << endl;

        cout << shadow->val;
        ListNode* temp = shadow->next;
        shadow->next = temp->next;
        delete temp;

        return head;
    }
};
